vital/algo/transform_io.h

```diff
@@ -0,0 +1,136 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/**
+ * \file
+ * \brief Interface for transform_io \link kwiver::vital::algo::algorithm_def
+ *        algorithm definition \endlink.
+ */
+
+#ifndef VITAL_ALGO_TRANSFORM_IO_H_
+#define VITAL_ALGO_TRANSFORM_IO_H_
+
+#include <vital/vital_config.h>
+
+#include <string>
+
+#include <vital/algo/algorithm.h>
+#include <vital/types/transform.h>
+
+namespace kwiver {
+namespace vital {
+namespace algo {
+
+/// An abstract base class for reading and writing transforms
+/**
+ * This class represents an abstract interface for reading and writing
+ * transforms.
+ */
+class VITAL_ALGO_EXPORT transform_io
+  : public kwiver::vital::algorithm_def<transform_io>
+{
+public:
+  virtual ~transform_io() = default;
+
+  /// Return the name of this algorithm
+  static std::string static_type_name() { return "transform_io"; }
+
+  /// Load transform from the file
+  /**
+   * \throws kwiver::vital::path_not_exists
+   *   Thrown when the given path does not exist.
+   *
+   * \throws kwiver::vital::path_not_a_file
+   *   Thrown when the given path does not point to a file (i.e. it points to a
+   *   directory).
+   *
+   * \param filename the path to the file to load
+   * \returns a transform instance referring to the loaded transform
+   */
+  kwiver::vital::transform_sptr load( std::string const& filename ) const;
+
+  /// Save transform to a file
+  /**
+   * Transform file format is based on the algorithm instance.
+   *
+   * \throws kwiver::vital::path_not_exists
+   *   Thrown when the expected containing directory of the given path does not
+   *   exist.
+   *
+   * \throws kwiver::vital::path_not_a_directory
+   *   Thrown when the expected containing directory of the given path is not
+   *   actually a directory.
+   *
+   * \throws kwiver::vital::invalid_data
+   *   Thrown when the algorithm does not recognize the concrete type of the
+   *   transformation instance.
+   *
+   * \param filename the path to the file to save
+   * \param data the transform instance referring to the transform to write
+   */
+  void save( std::string const& filename,
+             kwiver::vital::transform_sptr data ) const;
+
+protected:
+  transform_io();
+
+private:
+  /// Implementation specific load functionality.
+  /**
+   * Concrete implementations of transform_io class must provide an
+   * implementation for this method.
+   *
+   * \param filename the path to the file the load
+   * \returns a transform instance referring to the loaded transform
+   */
+  virtual kwiver::vital::transform_sptr load_(
+    std::string const& filename ) const = 0;
```

Comment:
Would load_ and save_ be more useful if they operated on a stream rather than a file name?

Reply:
Maybe, maybe not... that would give the flexibility to operate on something that is not a file, but only if the implementation is able to operate on a `std::istream`.

At any rate, this is copied from `image_io`...